### PR TITLE
Add support for Socket Logging Handler with basic ECS format logging

### LINKED ISSUE #23127: Logging into socket
### Description

Hello, 
I would like to add option to use plain UDP/TCP/SSL_TCP socket logging which is supported out of the box by JBossLogManager. In our environment we dont have currently allowed to use GELF nor Syslog. We were able to log from our app to logstash but those 2 input methods are not currently allowed.

We would like to add basic formatters like plain and JSON.
In Json: we would like to add custom fields to log message which can be setup 'statically' during launch time. Exactly as GELF with additional fields. We also need option to override keys as some fields are differently named in StructuredFormatter.

### Configuration suggestion

quarkus.log.handler.socket.enabled=true
quarkus.log.handler.socket.host=localhost
quarkus.log.handler.socket.port=12201
quarkus.log.handler.socket.protocol=UDP
quarkus.log.handler.socket.formatter=json
quarkus.log.handler.socket.keyoverrides=LEVEL=severity
quarkus.log.handler.socket.additional-field.serverName.value=myServerName

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by troosan**
This pull request is a followup of PR #23128 and is greatly inspired by the following provided by @geoand https://github.com/quarkusio/quarkus/compare/main...geoand:%2323127?expand=1

It allows sending logs in ECS format to a socket (typically a logstash tcp input)

fixes #23127 and answers my own question on [StackOverFlow](https://stackoverflow.com/questions/78952158/send-quarkus-logs-to-logstash-elk-in-json-format-over-logstash-tcp-input)

### CODE REVIEW

**Comment by troosan** [comment]
I'm still working on updating the documentation for it (logging and centralized-log-management)

**Comment by geoand** [comment]
Looks good, thanks!

@dmlloyd any commenta?

**Comment by troosan** [comment]
@dmlloyd @geoand should we natively provide an valid ECS json?
- using an external library (provided by [elastic](https://www.elastic.co/guide/en/ecs-logging/java/1.x/setup.html#_add_the_dependency)) It looks quite lightweight: https://mvnrepository.com/artifact/co.elastic.logging/jboss-logmanager-ecs-formatter/1.6.0
- by building the json manually

or should I update the documentation to mention the quarkiverse version of the quarkus-logging-json, which supports ECS encoding. This will need a PR for that extension though to enable using the socket handler.

**Comment by geoand** [comment]
@loicmathieu you've used the logging-json stuff IIRC, what's your take?

**Comment by loicmathieu** [comment]
`quarkus-logging-json` is the official way to log in JSON.

What I don't understand is why it mandates modification in the logging-json library, socket loggings should be a log handler so it should work with all supported log format/pattern?

By the way, `logging-gelf` already support that.

**Comment by troosan** [comment]
I think the reason this required a change in the logging-json is because it's configured in a different way.
The json logging can be configured per handler (console, file and now socket) whereas the gelf handler handles only it's own format.

I know the gelf logging already supports sending logs to logstash, but my operations team is not too keen on enabling & maintaining another input (and the filters that go with it) in the logstash pipeline.

**Comment by troosan** [comment]
@geoand @gsmet Contrary to what I put int the [documentation](https://github.com/quarkusio/quarkus/pull/43232/files#diff-0d1d184ad8a94ae64442b6f54113d08bd13a6cfe092cb911b069fda0762b9a18R517), with this current PR, the format of the JSON does not follow the [ECS](https://www.elastic.co/guide/en/ecs/current/ecs-guidelines.html) standards.

For now I have gone around this by setting field key overrides through the configuration, hiding/adding additional fields, like this:
```
quarkus.log.socket.json.exception-output-type=formatted
quarkus.log.socket.json.key-overrides=timestamp=@timestamp,\
	logger_name=log.logger,\
	level=log.level,\
	thread_name=process.thread.name,\
	thread_id=process.thread.id,\
	host_name=host.hostname,\
	stack_trace=error.stack_trace
quarkus.log.socket.json.excluded-keys=sequence,loggerClassName
quarkus.log.socket.json.additional-field."service.environment".value=${quarkus.profile}
quarkus.log.socket.json.additional-field."service.name".value=${quarkus.application.name}
quarkus.log.socket.json.additional-field."service.version".value=${quarkus.application.version}
quarkus.log.socket.json.additional-field."data_stream.type".value=logs
quarkus.log.socket.json.additional-field."ecs.version".value=1.12
```

This is only however not a very clean solution, I could also do this programmatically like this in the `LoggingJsonRecorder`
```java
        if (config.logFormat == JsonConfig.LogFormat.ECS) {
            EnumMap<Key, String> keyOverrides = PropertyValues.stringToEnumMap(Key.class, config.keyOverrides.orElse(null));
            keyOverrides.putIfAbsent(Key.TIMESTAMP, "@timestamp");
            keyOverrides.putIfAbsent(Key.LOGGER_NAME, "log.logger");
            keyOverrides.putIfAbsent(Key.LEVEL, "log.level");
            keyOverrides.putIfAbsent(Key.THREAD_NAME, "process.thread.name");
            keyOverrides.putIfAbsent(Key.THREAD_ID, "process.thread.id");
            keyOverrides.putIfAbsent(Key.HOST_NAME, "host.hostname");
            keyOverrides.putIfAbsent(Key.STACK_TRACE, "error.stack_trace");
            config.keyOverrides = Optional.of(PropertyValues.mapToString(keyOverrides));
        }
```
which is not very clean either.

Also missing is the addition of fields like `trace.id` and `span.id` instead (or in addition to) of having them in the `mdc.traceId` and `mdc.spanId`

So question is, shall I commit this working (but far from perfect) solution and adapt the documentation accordingly?
Or do I just specify in the documentation how to make it ECS compatible through properties?

**Comment by geoand** [comment]
> Or do I just specify in the documentation how to make it ECS compatible through properties?

I think this is best for now. @loicmathieu WDYT?

**Comment by troosan** [comment]
@geoand in case we go for the code solution instead of the properties solution, do you know how I can get the traceId and spanId? Those will obviously only be available if the OpenTelemetry dependency is present. Is there a way to retrieve those two values without relying on the `io.opentelemetry.context` classes?
I would like to add those in the [ECS tracing fields](https://www.elastic.co/guide/en/ecs/current/ecs-tracing.html)
My current solution is to copy them in the Logstash pipeline with the following filter but it would be nicer to natively support it.
```
filter {
  if ![span][id] and [mdc][spanId] {
    mutate { rename => { "[mdc][spanId]" => "[span][id]" } }
  }
  if ![trace][id] and [mdc][traceId] {
    mutate { rename => {"[mdc][traceId]" => "[trace][id]"} }
  }
}
```

**Comment by geoand** [comment]
> @geoand in case we go for the code solution instead of the properties solution, do you know how I can get the traceId and spanId? Those will obviously only be available if the OpenTelemetry dependency is present. Is there a way to retrieve those two values without relying on the io.opentelemetry.context classes?

@brunobat can give you the best way to access those, although in order to make reduce unnecessary couplinh, we'd need to put some kind of SPI in place.

**Comment by brunobat** [comment]
@troosan  You need access to the context like the MDCEnabledContextStorage does: https://github.com/quarkusio/quarkus/blob/bbd821aaf1208187249a80c133826121bdd27468/extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/MDCEnabledContextStorage.java#L30 

But this data is maintained by us in the MDC context, therefore, your solution sounds ok.

**Comment by loicmathieu** [comment]
I think we should have a util for that that can extract the spanId/transactionId if present as we need to do it for all handlers. For ex it would be useful to be able to forward these ids to Google Cloud Loggin, today it must be done manually by implementing a trace extractor: https://docs.quarkiverse.io/quarkus-google-cloud-services/main/logging.html#_tracing

This is a cross-extension concern so we should have something in the core that didn't depends on OTEL.

@troosan 
> Or do I just specify in the documentation how to make it ECS compatible through properties?
Yes, again we may want to provide something that will work cross handler, for example this is the code we have for Google Cloud Logging: https://github.com/quarkiverse/quarkus-google-cloud-services/blob/main/logging/runtime/src/main/java/io/quarkiverse/googlecloudservices/logging/runtime/ecs/EscJsonFormat.java

**Comment by troosan** [comment]
@loicmathieu that is indeed exactly what we need, Ideally the generation of the json (ECS or not) and where to send it (Google, Logstash, ...) should be decoupled. Otherwise we're just duplicating code all over.

**Comment by troosan** [comment]
@geoand I have updated the documentation on how to make the json ECS compatible through properties.
https://quarkus-pr-main-43232-preview.surge.sh/version/main/guides/centralized-log-management#logstash_ecs

Do you think this could get merged as is in next iteration? (I can squash the commits before merging)

Longer term the logging facility probably needs to be reviewed to avoid unnecessary duplication across the extensions.

File: core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java
Comment (gsmet) [inline core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java:61]: Is it expected that it references syslog?

File: docs/src/main/asciidoc/centralized-log-management.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/centralized-log-management.adoc:243]:
```suggestion
You can also send your logs to Logstash using a TCP input and the https://www.elastic.co/guide/en/ecs-logging/overview/current/intro.html[ECS] format.
```

File: docs/src/main/asciidoc/centralized-log-management.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/centralized-log-management.adoc:244]:
```suggestion
To achieve this, we will use the `quarkus-logging-json` extension to format the logs in JSON format and the socket handler to send them to Logstash.
```

File: docs/src/main/asciidoc/centralized-log-management.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/centralized-log-management.adoc:246]:
```suggestion
For this you can use the same `docker-compose.yml` file as above but with a different Logstash pipeline configuration.
```

File: docs/src/main/asciidoc/centralized-log-management.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/centralized-log-management.adoc:274]:
```suggestion
Then configure your application to log in JSON format instead of GELF:
```

File: docs/src/main/asciidoc/logging.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/logging.adoc:515]:
```suggestion
When enabled, it sends all log events to a socket, for instance to a Logstash server.
```

File: core/runtime/src/main/java/io/quarkus/runtime/logging/LogConfig.java
Comment (gsmet) [inline core/runtime/src/main/java/io/quarkus/runtime/logging/LogConfig.java:111]:
```suggestion
     * The named socket handlers configured here can be linked to one or more categories.
```

**Review by gsmet — CHANGES_REQUESTED** [review]
I added some minor suggestions and questions.

Other than that, I think the patch looks good and it's a very nice contribution.

File: core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java
Comment (dmlloyd) [inline core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java:13]: Sorry for the late review, but if this is being used specifically to talk to one kind of service, then should we be having a general socket config? I think that users might have  a harder time configuring a generic socket handler to talk to some service than if they could have a configuration specific to the kind of service they're talking to.

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/LogSocketFormatBuildItem.java
Comment (dmlloyd) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/LogSocketFormatBuildItem.java:16]: Won't this break if we have more than one socket handler configured for some reason?

File: core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java
Comment (troosan) [reply core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java:13]: Theoretically the socket to which you are writing is independent of the format in which you send the data to it.
The current setup will allow you to format the logs however you want, JSON is only one of those.

The Socket handler should allow you to send logs to [Splunk](https://dev.splunk.com/enterprise/docs/devtools/java/logging-java/howtouseloggingjava/enablelogtcpjava/) too for instance. I've not tried it out though.

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/LogSocketFormatBuildItem.java
Comment (troosan) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/LogSocketFormatBuildItem.java:16]: you mean that we should be able to work with "named" handlers? To allow writing logs to 2 different socket appenders? (Logstash and Splunk for instance)

I must admit I don't know. This is just a copy of `LogFileFormatBuildItem` for instance

File: core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java
Comment (troosan) [reply core/runtime/src/main/java/io/quarkus/runtime/logging/SocketConfig.java]: copy paste ...

**Comment by troosan** [comment]
> I added some minor suggestions and questions.
> 
> Other than that, I think the patch looks good and it's a very nice contribution.

Fixed what you reported, squashed and forced pushed again. For some reason one of the tests fails (native HTTP tests), can those be rerun?

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
@dmlloyd I think @troosan answered your concerns. From my POV, it's in line with what we already have for the other logging handler so I wouldn't block this one because of:

> Won't this break if we have more than one socket handler configured for some reason?

Let us know what you think.

**Comment by troosan** [comment]
> I don't really like the configuration. It seems over the top. I would instead rather be able to configure a handler for logstash or splunk, by name, and point it at a socket address and have that be the extent of the configuration necessary.

That would be another way to do it, but to be consistent you would need to change the whole way logging is configured right now. The elastic json log format (ECS) can be used for (using filebeat for file based appenders)
 - console
 - file
 - socket
If I understand correctly what you are suggesting you would prefer to configure first the format, then the handler, so `quarkus.logging.elastic.console` instead of `quarkus.logging.console.elastic`, which is the other way around than what is done now. That would be a major breaking change though ...
 
> I've lost this battle before in a big way (data source URLs...) so if you want to override my objection go ahead. I would just prefer to have fewer low level details in the configuration, because avoiding that stuff is supposed to be a part of the value we add.

**Comment by troosan** [comment]
I guess everybody was in Antwerp at Devoxx last week :-)
Any chance to have this part of the next iteration?

File: docs/src/main/asciidoc/centralized-log-management.adoc
Comment (dmlloyd) [inline docs/src/main/asciidoc/centralized-log-management.adoc:312]: This I just don't like. Could we at least have a enum property on the JSON formatter for format presets? Asking every user to provide all of this each time is not OK IMO.

Comment (troosan) [reply docs/src/main/asciidoc/centralized-log-management.adoc:312]: that's indeed what I proposed here: https://github.com/quarkusio/quarkus/pull/43232#issuecomment-2352186654

**Comment by troosan** [comment]
looks like the move of the logging to @ConfigMapping made a mess out of my PR. I'll update it.
@geoand @dmlloyd @gsmet what do you prefer concerning https://github.com/quarkusio/quarkus/pull/43232#discussion_r1801526051 ?

**Comment by troosan** [comment]
@dmlloyd what shall I do for this PR? I personally would tend to agree with you that it's much easier for the enduser to just specify quarkus.log.socket.json.log-format=ECS and do the key-overrides programmatically.
Once someone comes up with a "clean" way to produce the ECS formatted json, we can just swap the impl. The config key can stay the same, so it would be backward compatible.

**Comment by dmlloyd** [comment]
> @dmlloyd what shall I do for this PR? I personally would tend to agree with you that it's much easier for the enduser to just specify quarkus.log.socket.json.log-format=ECS and do the key-overrides programmatically. Once someone comes up with a "clean" way to produce the ECS formatted json, we can just swap the impl. The config key can stay the same, so it would be backward compatible.

Sure, I like this approach. That would cover my two concerns (configurability, plus being able to clean up things in the future).

**Comment by troosan** [comment]
@dmlloyd changes done and documentation updated accordingly

**Review by dmlloyd — APPROVED** [review]
Looks OK. I think we'll probably want to revisit some of these internals at some point, but you've really done your best to accommodate and I think the current state is probably acceptable for now. Thanks.

**Comment by troosan** [comment]
you're welcome. I really need to dive into the quarkus build specificities to understand better how all this works !

**Comment by cescoffier** [comment]
@loicmathieu @dmlloyd Are we good on this one?

**Comment by dmlloyd** [comment]
I'm good with it

**Comment by cescoffier** [comment]
Thanks @dmlloyd !

**Review by loicmathieu — APPROVED** [review]

**Comment by loicmathieu** [comment]
Good for me.

I'll open a discussion so review all ou centralized log solutions now that we have 3 of them: socket appender, logstash gelf and Open telemetry logs we would want to document the differences and maybe deprecate the old logstash gelf.

I'm also have a look if I can reuse the new ECS formatter in Google Cloud Logging

**Pull request merged**
